In [2]:
import pandas as pd
url = "https://drive.google.com/uc?export=download&id=1eJ-L0MnJxHi929fZrVOKxH7Mrw5eTGpk"
df = pd.read_csv(url)

/tmp/ipykernel_15011/2739466525.py:3: DtypeWarning: Columns (0,3,5) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(url)


# Atividade 1 — Exercícios com os comandos básicos de pandas para EDA


Responda as perguntas abaixo no notebook (cada resposta deve ter ao menos uma célula de código que produza o output e, logo abaixo, uma célula de texto com a sua interpretação).

## 1) Carregue o dataset e mostre as primeiras 10 linhas.

In [3]:
df.head(10)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"700000,00","0,00","0,00","0,00","690900,00","9100,00"
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"487500,00","0,00","0,00","0,00","487500,00","0,00"
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"100000,00","0,00","0,00","0,00","0,00","100000,00"
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","243750,00","0,00"
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"1658333,94","0,00","0,00","0,00","527200,00","1131133,94"
5,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CAMPINAS DO SUL - RS,4303806,CAMPINAS DO SUL,4300000,...,10V0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0000,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,"243750,00","0,00","0,00","0,00","0,00","243750,00"
6,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,MORENO - PE,2609402,MORENO,2600000,...,20RP,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0000,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,"700000,00","0,00","0,00","0,00","700000,00","0,00"
7,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,NOVA FLORESTA - PB,2510105,NOVA FLORESTA,2500000,...,8581,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0000,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,"278700,00","0,00","0,00","0,00","0,00","278700,00"
8,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,SANTIAGO DO SUL - SC,4215695,SANTIAGO DO SUL,4200000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"300000,00","300000,00","300000,00","0,00","0,00","0,00"
9,Sem informação,2014,Emenda Individual - Transferências com Finalid...,S/I,Sem informação,S/I,CIANORTE - PR,4105508,CIANORTE,4100000,...,1D73,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0000,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,"245850,00","0,00","0,00","0,00","0,00","245850,00"


Podemos obsevar uma tabela contendo várias informações, de vários tipos.

## 2) Quantas linhas e colunas o dataset possui?

In [4]:
df.shape

(88596, 28)

Basicamente a função "df.shape" mostra a quantidade de linhas e colunas, no qual o primeiro número "88596" representa a quantidade de linhas, e o segundo número "28" representa a quantidade de colunas.

## 3) Liste os nomes das colunas e descreva brevemente o que você entende que cada coluna representa.

In [5]:
df.columns

Index(['Código da Emenda', 'Ano da Emenda', 'Tipo de Emenda',
       'Código do Autor da Emenda', 'Nome do Autor da Emenda',
       'Número da emenda', 'Localidade de aplicação do recurso',
       'Código Município IBGE', 'Município', 'Código UF IBGE', 'UF', 'Região',
       'Código Função', 'Nome Função', 'Código Subfunção', 'Nome Subfunção',
       'Código Programa', 'Nome Programa', 'Código Ação', 'Nome Ação',
       'Código Plano Orçamentário', 'Nome Plano Orçamentário',
       'Valor Empenhado', 'Valor Liquidado', 'Valor Pago',
       'Valor Restos A Pagar Inscritos', 'Valor Restos A Pagar Cancelados',
       'Valor Restos A Pagar Pagos'],
      dtype='object')

Basicamente, podemos observar que existem 28 colunas contendo informções únicas, sendo cada:

• Código da Emenda: Identificador único da
emenda parlamentar.

• 	Ano da Emenda: Ano em que a emenda foi apresentada ou aprovada.

• 	Tipo de Emenda: Classificação da emenda (individual, de bancada, de comissão etc.).

• 	Código do Autor da Emenda: Identificador numérico do autor da emenda.

• 	Nome do Autor da Emenda: Nome completo do parlamentar ou entidade que propôs a emenda.

• 	Número da Emenda: Número sequencial atribuído à emenda.

• 	Localidade de aplicação do recurso: Local onde os recursos da emenda serão aplicados.

• 	Código Município IBGE: Código oficial do município segundo o IBGE.

• 	Município: Nome do município beneficiado.

• 	Código UF IBGE: Código da Unidade Federativa segundo o IBGE.

• 	UF: Sigla da Unidade Federativa (estado).

• 	Região: Região geográfica do Brasil (Norte, Nordeste, Centro-Oeste, Sudeste, Sul).

• 	Código Função: Código da função orçamentária (ex.: Saúde, Educação).

• 	Nome Função: Nome da função orçamentária correspondente.

• 	Código Subfunção: Código da subfunção dentro da função orçamentária.

• 	Nome Subfunção: Nome da subfunção correspondente.

• 	Código Programa: Código do programa de governo vinculado à emenda.

• 	Nome Programa: Nome do programa de governo.

• 	Código Ação: Código da ação orçamentária específica.

• 	Nome Ação: Nome da ação orçamentária correspondente.

• 	Código Plano Orçamentário: Código do plano orçamentário vinculado.

• 	Nome Plano Orçamentário: Nome do plano orçamentário correspondente.

• 	Valor Empenhado: Valor autorizado e reservado para execução da despesa.

• 	Valor Liquidado: Valor da despesa efetivamente realizada e reconhecida.

• 	Valor Pago: Valor efetivamente pago ao fornecedor ou beneficiário.

• 	Valor Restos A Pagar Inscritos: Valores de despesas empenhadas, mas não pagas até o fim do exercício.

• 	Valor Restos A Pagar Cancelados: Valores de restos a pagar que foram
anulados.

• 	Valor Restos A Pagar Pagos: Valores de restos a pagar que foram efetivamente quitados.

## 4) Existem valores nulos? quantos?


In [6]:
df.isnull().sum()

,0
Código da Emenda,0
Ano da Emenda,0
Tipo de Emenda,0
Código do Autor da Emenda,0
Nome do Autor da Emenda,0
Número da emenda,0
Localidade de aplicação do recurso,0
Código Município IBGE,0
Município,0
Código UF IBGE,0


Como podemos observar, existem um total de 0 - zero valores nulos, e em todas as colunas.

## 5) Mostre os tipos das variáveis (se int, string, boolean, float...) e analise se alguma variável parece ter o tipo errado. Faça isso da maneira mais objetiva e otimizada possível.

In [7]:
df.dtypes

,0
Código da Emenda,object
Ano da Emenda,int64
Tipo de Emenda,object
Código do Autor da Emenda,object
Nome do Autor da Emenda,object
Número da emenda,object
Localidade de aplicação do recurso,object
Código Município IBGE,object
Município,object
Código UF IBGE,int64


Se alguma coluna estiver com o tipo errado, escreva abaixo qual é e por que isso pode atrapalhar análises.

**Dica:** Se alguma variável tiver no tipo o nome "object", isso se refere a texto.

Como podemos observar, temos algumas variáveis que estão com um tipo de variável incorreta da qual deveria estar. As colunas de valores financeiros estão com um tipo incorreto " int64 ou object ", quando deveriam ser "float64". Isso atrapalha porque impede cálculos precisos e análises estatísticas corretas, já que números tratados como texto não podem ser somados ou comparados

## 6) Usando a coluna `Ano da Emenda` como base, descubra até que ano o dataset está atualizado (mostre as 8 primeiras linhas do df ordenado desc).

In [8]:
df.sort_values(by = "Ano da Emenda", ascending = False).head(8)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
88564,202541820003,2025,Emenda Individual - Transferências Especiais,4182,ZEQUINHA MARINHO,3,MÚLTIPLO,Sem informação,Múltiplo,1500000,...,0EC2,TRANSFERENCIAS ESPECIAIS,0000,TRANSFERENCIAS ESPECIAIS,"12161000,00","12039390,00","12039390,00","0,00","0,00","0,00"
88578,202544940001,2025,Emenda Individual - Transferências com Finalid...,4494,ZEZINHO BARBARY,1,Nacional,Sem informação,Sem informação,-1,...,2726,PREVENCAO E REPRESSAO AO TRAFICO ILICITO DE DR...,0000,PREVENCAO E REPRESSAO AO TRAFICO ILICITO DE DR...,"478233,30","0,00","0,00","0,00","0,00","0,00"
88595,202544840009,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,9,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,20JP,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,0000,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,"1800000,00","36000,00","0,00","0,00","0,00","0,00"
88594,202544840008,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,8,Nacional,Sem informação,Sem informação,-1,...,2723,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...",0000,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...","497151,34","0,00","0,00","0,00","0,00","0,00"
88577,202541820017,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,17,REDENÇÃO - PA,1506138,REDENÇÃO,1500000,...,219G,ESTRUTURACAO DA REDE DE SERVICOS E FORTALECIME...,0000,ESTRUTURACAO DA REDE DE SERVICOS E FORTALECIME...,"370000,00","370000,00","370000,00","0,00","0,00","0,00"
88576,202541820016,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,16,PARÁ (UF),Sem informação,Sem informação,1500000,...,8535,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,0000,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,"600000,00","0,00","0,00","0,00","0,00","0,00"
88575,202541820015,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,15,PARÁ (UF),Sem informação,Sem informação,1500000,...,20Z1,QUALIFICACAO SOCIAL E PROFISSIONAL DE TRABALHA...,0000,QUALIFICACAO SOCIAL E PROFISSIONAL DE TRABALHA...,"500000,00","0,00","0,00","0,00","0,00","0,00"
88574,202541820014,2025,Emenda Individual - Transferências com Finalid...,4182,ZEQUINHA MARINHO,14,MÚLTIPLO,Sem informação,Múltiplo,1500000,...,00SL,APOIO A IMPLANTACAO E MODERNIZACAO DE INFRAEST...,0000,APOIO A IMPLANTACAO E MODERNIZACAO DE INFRAEST...,"500000,00","0,00","0,00","0,00","0,00","0,00"


Explique: qual é o ano mais recente presente no dataset?

O ano mais recente como podemos observar no dataset é: 2025

## 7) execute um  `describe()` e explique uma conclusão simples que você consegue tirar a partir das estatísticas (se algumas colunas deveriam aparecer no describe mas não aparecem, indique o motivo que você acredita que esteja causando esse problema).

In [10]:
df.describe()

,Ano da Emenda,Código UF IBGE
count,88596.000000,8.859600e+04
mean,2019.155109,2.943594e+06
std,3.409323,1.264884e+06
min,2014.000000,-1.400000e+01
25%,2016.000000,2.300000e+06
50%,2019.000000,3.100000e+06
75%,2022.000000,3.500000e+06
max,2025.000000,5.300000e+06


Como podemos observar, a função "describe()" mostra qual o intervalo de anos e tamém sua destribuição, nisso consiguimos observar uma variabilidade de dados. E algumas colunas não aparecem pois está função só permite analisar colunas com dados numéricos, e não textuais.

## 8) Mostre as últimas 5 linhas do dataset e escreva se a última linha parece completa (sem muitos nulos) ou não.

In [16]:
df.tail(5)

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Código Ação,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos
88591,202544840005,2025,Emenda Individual - Transferências Especiais,4484,ZUCCO,5,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,0EC2,TRANSFERENCIAS ESPECIAIS,0000,TRANSFERENCIAS ESPECIAIS,"12307991,08","12184912,08","12184912,08","0,00","0,00","0,00"
88592,202544840006,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,6,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,20ZV,FOMENTO AO SETOR AGROPECUARIO,0000,FOMENTO AO SETOR AGROPECUARIO - DESPESAS DIVERSAS,"988649,90","25592,00","25592,00","0,00","0,00","0,00"
88593,202544840007,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,7,Nacional,Sem informação,Sem informação,-1,...,2000,ADMINISTRACAO DA UNIDADE,0000,ADMINISTRACAO DA UNIDADE - DESPESAS DIVERSAS,"999914,73","155246,57","149728,52","0,00","0,00","0,00"
88594,202544840008,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,8,Nacional,Sem informação,Sem informação,-1,...,2723,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...",0000,"POLICIAMENTO, FISCALIZACAO, ENFRENTAMENTO A CR...","497151,34","0,00","0,00","0,00","0,00","0,00"
88595,202544840009,2025,Emenda Individual - Transferências com Finalid...,4484,ZUCCO,9,MÚLTIPLO,Sem informação,Múltiplo,4300000,...,20JP,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,0000,DESENVOLVIMENTO DE ATIVIDADES E APOIO A PROGRA...,"1800000,00","36000,00","0,00","0,00","0,00","0,00"


In [17]:
ultima_linha = df.tail(1)
print(ultima_linha.isnull().sum())

Código da Emenda                      0
Ano da Emenda                         0
Tipo de Emenda                        0
Código do Autor da Emenda             0
Nome do Autor da Emenda               0
Número da emenda                      0
Localidade de aplicação do recurso    0
Código Município IBGE                 0
Município                             0
Código UF IBGE                        0
UF                                    0
Região                                0
Código Função                         0
Nome Função                           0
Código Subfunção                      0
Nome Subfunção                        0
Código Programa                       0
Nome Programa                         0
Código Ação                           0
Nome Ação                             0
Código Plano Orçamentário             0
Nome Plano Orçamentário               0
Valor Empenhado                       0
Valor Liquidado                       0
Valor Pago                            0


Como podemos observar, a última linha não contém dados nulos, e a última linha está completa, ou seja, contém dados válidos em praticamente todas as colunas.

## 9) existem valores duplicados no dataframe? Caso positivo, quantos?

In [14]:
duplicados = df.duplicated().sum()
print(f'O número de linhas duplicadas é: {duplicados}')

O número de linhas duplicadas é: 0


Como podemos observar através da função, não existem valores duplicados.